crates/codspeed/src/fifo.rs

```diff
@@ -0,0 +1,201 @@
+pub const RUNNER_ACK_FIFO: &str = "/tmp/runner.ack.fifo";
+
+pub struct PerfGuard {
+    ctl_fifo: FifoIpc,
+    ack_fifo: FifoIpc,
+}
+
+impl PerfGuard {
+    pub fn new(ctl_fifo: &str, ack_fifo: &str) -> Option<Self> {
+        let mut instance = Self {
+            ctl_fifo: FifoIpc::connect(ctl_fifo)?.with_writer().ok()?,
+            ack_fifo: FifoIpc::connect(ack_fifo)?.with_reader().ok()?,
+        };
+        instance.send_cmd(Command::StartBenchmark)?;
+        Some(instance)
+    }
+
+    fn send_cmd(&mut self, cmd: Command) -> Option<()> {
+        self.ctl_fifo.send_cmd(cmd)?;
+        self.ack_fifo.wait_for_ack()?;
+
+        Some(())
+    }
+}
+
+impl Drop for PerfGuard {
+    fn drop(&mut self) {
+        self.send_cmd(Command::StopBenchmark);
+    }
+}
+
+pub struct FifoIpc {
+    path: PathBuf,
+    reader: Option<File>,
+    writer: Option<File>,
+}
+
+impl FifoIpc {
+    pub fn connect<P: Into<PathBuf>>(path: P) -> Option<Self> {
+        let path = path.into();
+
+        if !path.exists() {
+            return None;
+        }
+
+        Some(Self {
+            path,
+            reader: None,
+            writer: None,
+        })
+    }
+
+    pub fn create(path: &str) -> Option<Self> {
+        // Remove the previous FIFO (if it exists)
+        let _ = unlink(path);
+
+        // Create the FIFO with RWX permissions for the owner
+        unistd::mkfifo(path, stat::Mode::S_IRWXU).unwrap();
+
+        Self::connect(path)
+    }
+
+    pub fn with_reader(mut self) -> std::io::Result<Self> {
+        self.reader = Some(
+            OpenOptions::new()
+                .write(true)
+                .read(true)
+                .custom_flags(O_NONBLOCK)
+                .open(&self.path)?,
+        );
+        Ok(self)
+    }
+
+    /// WARNING: Writer must be opened _AFTER_ the reader.
+    pub fn with_writer(mut self) -> std::io::Result<Self> {
+        self.writer = Some(
+            OpenOptions::new()
+                .write(true)
+                .custom_flags(O_NONBLOCK)
+                .open(&self.path)?,
+        );
+        Ok(self)
+    }
+
+    pub fn recv_cmd(&mut self) -> Option<Command> {
+        // First read the length (u32 = 4 bytes)
+        let mut len_buffer = [0u8; 4];
+        self.read_exact(&mut len_buffer).ok()?;
+        let message_len = u32::from_le_bytes(len_buffer) as usize;
+
+        // Try to read the message
+        let mut buffer = vec![0u8; message_len];
+        loop {
+            if self.read_exact(&mut buffer).is_ok() {
+                break;
+            }
+        }
+
+        let decoded = bincode::deserialize(&buffer).ok()?;
+        Some(decoded)
+    }
+
+    pub fn send_cmd(&mut self, cmd: Command) -> Option<()> {
+        let encoded = bincode::serialize(&cmd).ok()?;
+        self.write_all(&(encoded.len() as u32).to_le_bytes()).ok()?;
+        self.write_all(&encoded).ok()?;
+        Some(())
+    }
+
+    pub fn wait_for_ack(&mut self) -> Option<()> {
```

Comment:
@art049 seemed keen on the idea of using `async` in this whole flow. I'm not very familiar with async so maybe he could give more arguments as to how it would benefit us

Reply:
Yes, overall, fo those functions I think it would make sense to use them asynchronously as much as possible to avoid polling. In the runner it should be possible everywhere

Reply:
But here, in the integrations, it might not be possible(we won't spin up an entire event loop just for this)